storage/index.go

```diff
@@ -130,42 +240,178 @@ func (i *nsIndex) Write(
+// NB: this function is called by the namespaceIndexInsertQueue.
+// FOLLOWUP(prateek): propagate error back up from here to the indexInsertQueue
+// so that we can notify users of success/failure correctly in the case of
+// sync'd inserts.
+func (i *nsIndex) writeBatch(inserts []index.WriteBatchEntry) {
 	// NB(prateek): we use a read lock to guard against mutation of the
-	// nsIndexBlock currently active, mutations within the underlying
-	// nsIndexBlock are guarded by primitives internal to it.
+	// indexBlocks, mutations within the underlying blocks are guarded
+	// by primitives internal to it.
 	i.RLock()
 	defer i.RUnlock()
 
 	if !i.isOpenWithRLock() {
 		// NB(prateek): deliberately skip calling any of the `OnIndexFinalize` methods
 		// on the provided inserts to terminate quicker during shutdown.
 		i.metrics.InsertAfterClose.Inc(int64(len(inserts)))
-		return errDbIndexUnableToWriteClosed
 	}
 
-	// NB: index.Block assumes responsibility for calling all the OnIndexSeries methods
-	// when WriteBatch is called up on it. Both on success and on failure.
-	_, err := i.active.WriteBatch(inserts)
-	return err
+	// we sort the inserts by which block they're applicable for, and do the inserts
+	// for each block.
+	writesByBlockStart := index.WriteBatchEntryByBlockStart(inserts)
+	sort.Sort(writesByBlockStart)
+	writesByBlockStart.ForEachBlockStart(i.writeBatchForBlockStartWithRLock)
+}
+
+func (i *nsIndex) writeBatchForBlockStartWithRLock(
+	blockStart xtime.UnixNano, inserts []index.WriteBatchEntry,
+) {
+	// ensure we have an index block for the specified blockStart.
+	block, err := i.ensureBlockPresentWithRLock(blockStart.ToTime())
+	if err != nil {
+		for _, insert := range inserts {
+			insert.OnIndexSeries.OnIndexFinalize()
+		}
+		i.logger.WithFields(
+			xlog.NewField("blockStart", blockStart),
+			xlog.NewField("numWrites", len(inserts)),
+			xlog.NewField("err", err.Error()),
+		).Error("unable to write to index, dropping inserts.")
+		i.metrics.AsyncInsertErrors.Inc(int64(len(inserts)))
+		return
+	}
+
+	// i.e. we have the block and the inserts, perform the writes.
+	result, err := block.WriteBatch(inserts)
+	// NB: we don't need to do anything to the OnIndexSeries refs in `inserts` at this point,
+	// the index.Block WriteBatch assumes responsibility for calling the appropriate methods.
+	if numErr := result.NumError; numErr != 0 {
+		i.metrics.AsyncInsertErrors.Inc(numErr)
+	}
+	if err != nil {
+		i.logger.Errorf("unable to write to index, dropping inserts. [%v]", err)
+	}
+}
+
+// Bootstrap bootstraps the index with the provide blocks.
+func (i *nsIndex) Bootstrap(
+	segmentsByBlockStart map[xtime.UnixNano][]segment.Segment,
+) error {
+	i.Lock()
+	if i.bootstrapState == Bootstrapping {
+		i.Unlock()
+		return errDbIndexIsBootstrapping
+	}
+	i.bootstrapState = Bootstrapping
+	i.Unlock()
+
+	defer func() {
+		i.Lock()
+		i.bootstrapState = Bootstrapped
+		i.Unlock()
+	}()
+
+	i.RLock()
+	defer i.RUnlock()
```

Comment:
Is this guaranteed to be called before the defer above?  Maybe it's better to:

Reply:
Yep Defers are guaranteed to be called in [reverse order of invocation](https://tour.golang.org/flowcontrol/13).

I thought about doing it that way originally but thought this approach was easier to follow. Will make the change if you find it easier to read.